ipa-core/src/protocol/context/dzkp_field.rs

```diff
@@ -12,17 +25,332 @@ pub trait DZKPCompatibleField: Field {
+                            // h3=b(1-2f),
+                            Fp61BitPrime::from_bit(*b) * one_minus_two_f,
+                            // h4=1-2f,
+                            one_minus_two_f,
+                        ]
+                    })
+                    .collect::<GenericArray<Fp61BitPrime, InitialRecursionFactor>>()
+            })
+    }
+}
+
+#[cfg(all(test, unit_test))]
+mod tests {
+    use bitvec::{array::BitArray, slice::BitSlice};
+    use rand::{thread_rng, Rng};
+
+    use crate::{
+        ff::Fp61BitPrime,
+        protocol::context::dzkp_field::{DZKPBaseField, SingleUVPolynomial, UVPolynomials},
+    };
+
+    #[test]
+    fn batch_convert() {
+        let mut rng = thread_rng();
+
+        // bitvecs
+        let mut vec_x_left = Vec::<u8>::new();
+        let mut vec_x_right = Vec::<u8>::new();
+        let mut vec_y_left = Vec::<u8>::new();
+        let mut vec_y_right = Vec::<u8>::new();
+        let mut vec_prss_left = Vec::<u8>::new();
+        let mut vec_prss_right = Vec::<u8>::new();
+        let mut vec_z_right = Vec::<u8>::new();
+
+        // gen 32 random values
+        for _i in 0..32 {
+            let x_left: u8 = rng.gen();
+            let x_right: u8 = rng.gen();
+            let y_left: u8 = rng.gen();
+            let y_right: u8 = rng.gen();
+            let prss_left: u8 = rng.gen();
+            let prss_right: u8 = rng.gen();
+            // we set this up to be equal to z_right for this local test
+            // local here means that only a single party is involved
+            // and we just test this against this single party
+            let z_right: u8 = (x_left & y_left)
+                ^ (x_left & y_right)
+                ^ (x_right & y_left)
+                ^ prss_left
+                ^ prss_right;
+
+            // fill vec
+            vec_x_left.push(x_left);
+            vec_x_right.push(x_right);
+            vec_y_left.push(y_left);
+            vec_y_right.push(y_right);
+            vec_prss_left.push(prss_left);
+            vec_prss_right.push(prss_right);
+            vec_z_right.push(z_right);
+        }
+
+        // conv to BitVec
+        let x_left = BitArray::<[u8; 32]>::try_from(BitSlice::from_slice(&vec_x_left)).unwrap();
+        let x_right = BitArray::<[u8; 32]>::try_from(BitSlice::from_slice(&vec_x_right)).unwrap();
+        let y_left = BitArray::<[u8; 32]>::try_from(BitSlice::from_slice(&vec_y_left)).unwrap();
+        let y_right = BitArray::<[u8; 32]>::try_from(BitSlice::from_slice(&vec_y_right)).unwrap();
+        let prss_left =
+            BitArray::<[u8; 32]>::try_from(BitSlice::from_slice(&vec_prss_left)).unwrap();
+        let prss_right =
+            BitArray::<[u8; 32]>::try_from(BitSlice::from_slice(&vec_prss_right)).unwrap();
+        let z_right = BitArray::<[u8; 32]>::try_from(BitSlice::from_slice(&vec_z_right)).unwrap();
+
+        // check correctness of the polynomials
```

Comment:
This test only tests consistency, not correctness. All we learn from this is that prover and verifier and computing these things in the same way.

It would be good to also have a test for correctness. I'd recommend grabbing some sample values from the excel spreadsheet to make sure they all match up.

Reply:
which excel spreadsheet are you referring to? How can we be sure that those example values are correct? I will add a test for the prover's bits. Since the provers bit are tested for consistency with the verifiers bits, the correctness of the verifiers bits follows then.

Reply:
I realised there is a problem: The excel spreadsheet cannot work with our 61-bit prime. It doesn't have that level of precision.